### PR TITLE
Add byte size limit option to ElasticsearchIO writer (#1829)

### PR DESCRIPTION
**Pull request by alexclare**
Adds the ability to split up bulk Elasticsearch writes by request byte size in addition to request count.

Sort of a prerequisite for #1938. Not strictly required, but that issue involves larger Java changes so I figured it would be better to get this in beforehand.

### CODE REVIEW

File: scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 1]
```diff
@@ -152,6 +160,7 @@
     public static class Bound<T> extends PTransform<PCollection<T>, PDone> {
 
       private static final int CHUNK_SIZE = 3000;
+      private static final long CHUNK_BYTES = 5L * 1024L * 1024L;
```
Comment (regadas) [inline scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:163]: Can you add a small doc line about this value?

File: scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 2]
```diff
@@ -415,32 +456,47 @@ public void setup() throws Exception {
...
         final Stream<DocWriteRequest> docWriteRequests =
             StreamSupport.stream(values.spliterator(), false)
                 .map(toDocWriteRequests::apply)
                 .flatMap(ar -> StreamSupport.stream(ar.spliterator(), false));
 
-        final Iterable<List<DocWriteRequest>> chunks =
-            Iterables.partition(docWriteRequests::iterator, maxBulkRequestSize);
+        int currentSize = 0;
+        long currentBytes = 0L;
+        ImmutableList.Builder<DocWriteRequest> chunk = ImmutableList.builder();
```
Comment (regadas) [inline scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:466]: I think `ImmutableList` is a bit unnecessary here.

File: scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java [hunk 3]
```diff
@@ -415,32 +456,47 @@ public void setup() throws Exception {
       public void processElement(ProcessContext c) throws Exception {
         final Iterable<T> values = c.element().getValue();
 
-        // Elasticsearch throws ActionRequestValidationException if bulk request is empty,
-        // so do nothing if number of actions is zero.
-        if (!values.iterator().hasNext()) {
```
Comment (regadas) [inline scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:420]: we can still do an early get out here and avoid the allocations bellow, right?

**Review by regadas — COMMENTED** [review]
Thank you for this @alexclare! just some minor comments!

File: scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java (hunk 3 above)
Comment (alexclare) [reply scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:420]: Yeah, that's true. Should I leave this block in then?

Comment (regadas) [reply scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java:420]: yeah let's keep it.

File: scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java (hunk 2 above)
Comment (alexclare) [reply scio-elasticsearch/es6/src/main/java/org/apache/beam/sdk/io/elasticsearch/ElasticsearchIO.java]: I ended up just constructing the `BulkRequest` directly,

**Review by regadas — APPROVED** [review]
@alexclare Thanks for pushing this through!

**Pull request merged**